designs/node-upgrades.md

```diff
@@ -0,0 +1,187 @@
+
+### Option 1 - Upgrade Serially
+
+Karpenter will only upgrade one node at a time. Each node's upgrade will be complete when its replacement is ready and the old node is terminated. This naturally rate-limits node upgrades to node initialization time and means that users will only have to worry about one node disruption at a time with upgrades.
+
+With serial upgrades, new nodes that cannot become ready or old nodes that cannot fully drain can stop the upgrade process altogether. For example, an old node could fail to fully drain due to a PDB or a do-not-evict annotation, which may require manual intervention. A new node could fail to become ready for many reasons. For one, if the new AMI is incompatible with the cluster's applications, Karpenter should not upgrade more nodes as more upgrades to that AMI would likely fail. On the other hand, sometimes a failed upgrade can succeed in other cases like with an [Outdated CNI](https://karpenter.sh/v0.16.3/troubleshooting/#nodes-stuck-in-pending-and-not-running-the-kubelet-due-to-outdated-cni) - where using a newer instance type could fail where an older instance type would succeed.
+
+In order to not permanently block serial upgrades, upgrades will have timeouts. Relying on Kubernetes native concepts, Karpenter will compute a drain timeout that respects the `GracefulTerminationPeriodSeconds` on the pods on the nodes by summing all remaining `GracefulTerminationPeriodSeconds` fields. This accounts for the worst case where Karpenter can only evict one pod at a time, taking as long as possible. If Karpenter fails to create a replacement node, Karpenter can terminate the unitialized upgraded node, where rolling back is already complete.
+
+### Option 2 - Upgrade in Parallel
+
+For some use cases, upgrading serially may be too slow. To combat this, Karpenter can upgrade in parallel, allowing more than one node to upgrade at a time. In contrast to upgrading serially,  upgrading in parallel requires opinionated hyper-parameters or allowing an API for users to configure them.
+
+#### Option 2a - In parallel with controls
+
+Starting with an API, Karpenter could implement a Node Disruption Budget. Similar to PDBs, NDBs rate-limit the termination process of nodes, gating termination requests from the controller or users. Similar to a restrictive PDB, a restrictive NDB can be used to stop termination altogether. Upgrading serially and using a NDB with maxUnavailable to 1 are equivalent.
+
+```
+apiVersion: karpenter.sh/v1alpha1
+kind: NodeDisruptionBudget
+metadata:
+  name: default
+spec:
+  # Merged labelSelectors for multiple provisioners
+  labelSelectors:
```

Comment:
Maybe use the more expressive `matchExpression` syntax?

Reply:
That's a good idea! Since this section is explicitly called out as not the suggestion for now, I was thinking that we can come back to this in the future, if the current design is not good enough for users.